model/atmosphere/dycore/docs/icon4py_sphinx.py

```diff
@@ -0,0 +1,364 @@
+import types, typing
+import ast
+
+class FullMethodDocumenter(autodoc.MethodDocumenter):
+    """
+    'Fully' document a method, i.e. picking up and processing all 'tagged'
+    comment blocks in its source code.
+    """
+
+    objtype = 'full'
+    priority = autodoc.MethodDocumenter.priority - 1
+
+    # Configuration options
+    docstring_keyword = 'scidoc'
+    var_type_in_inputs = True
+    var_type_formatting = '``'
+    print_variable_longnames = True
+
+    def get_doc(self):
+        # Override the default get_doc method to pick up all docstrings in the
+        # source code
+        source = inspect.getsource(self.object) # this is only the source of the method, not the whole file
+
+        docstrings = self.get_docstring_blocks(source, self.docstring_keyword)
+
+        docstrings_list = []
+        for idocstr, docstring in enumerate(docstrings):
+            formatted_docstr = None
+
+            # Get some useful information on the following method call
+            call_string = self.get_next_method_call(source, source.find(docstring) + len(docstring))
+            next_method_info = self.get_method_info(call_string)
+            # and process a scientific documentation docstring
+            formatted_docstr = docstring.splitlines()
+            formatted_docstr = self.format_docstring_block(formatted_docstr, self.docstring_keyword)
+            formatted_docstr = self.add_header(formatted_docstr, next_method_info)
+            formatted_docstr = self.process_scidocstrlines(formatted_docstr, next_method_info)
+            formatted_docstr = self.add_next_method_call(formatted_docstr, call_string)
+
+            if formatted_docstr is not None:
+                if idocstr < len(docstrings)-1: # Add footer
+                    formatted_docstr = self.add_footer(formatted_docstr)
+                # add the processed docstring to the list
+                docstrings_list.append(formatted_docstr)
+
+        return docstrings_list
+    
+    def get_docstring_blocks(self, source, keyword):
+        # which in this implementation are comment blocks
+        comment_blocks = []
+        source_lines = source.splitlines()
+        in_block = False
+        current_block = []
+        for line in source_lines:
+            stripped_line = line.strip()
+            if stripped_line.startswith(f'# {keyword}:'):
+                if in_block:
+                    comment_blocks.append('\n'.join(current_block))
+                    current_block = []
+                in_block = True
+                current_block.append(line)
+            elif in_block:
+                if stripped_line.startswith('#'):
+                    current_block.append(line)
+                else:
+                    in_block = False
+                    comment_blocks.append('\n'.join(current_block))
+                    current_block = []
+        if current_block:
+            comment_blocks.append('\n'.join(current_block))
+        return comment_blocks
```

Comment:
I think the code can be simplified by using the last element of the `comment_blocks` list as current block and avoid storing the non useful comment marks of the scidoc which need to be removed later. I haven't tested the code at all, but I think something along these lines should work:
```suggestion
        comment_blocks = []
        in_block = False
        for line in source.splitlines():
            stripped_line = line.strip()
            is_comment = stripped_line.startswith('#')
            if stripped_line.startswith(f'# {keyword}:'):
                comment_blocks.append([])
            elif in_block and is_comment:
                comment_blocks[-1].append(f"{stripped_line}\n")
            in_block = is_comment

        return comment_blocks
```

By the way, the suggested code will not support the case where a new scidoc is started in the middle of another one. I think that case should be considered as an error since two different scidocs cannot document the same stencil call.
```
# scidoc:
# Outputs:
#  - z_exner_ic :
# scidoc:
# Outputs:
#  - z_exner_ic :
foo_stencil(...)
```

Reply:
I agree with most everything (and adapted), the only thing I kept is the entire block unformatted because I use it in its entirety when looking for the next method call:
`call_string = self.get_next_method_call(source, source.find(docstring) + len(docstring))`
I could re-write such that source.splitlines() is the "main processed object" and only `get_next_method_call` uses `''.join(source_splitlines)`, then base the start of the search on a line index instead of a char index as is now.
Let me know if you think it makes more sense.
(this option would also allow to clean up the lines directly when adding them to the block and avoid postprocessing them in a second step)